managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -3289,39 +3315,34 @@ public void setPropertiesFailed(ManagedLedgerException exception, Object ctx) {
         latch.await();
     }
 
-    @Override
-    public void asyncSetProperties(Map<String, String> properties, final SetPropertiesCallback callback, Object ctx) {
-        store.getManagedLedgerInfo(name, false, new MetaStoreCallback<ManagedLedgerInfo>() {
+    private void asyncUpdateProperties(Map<String, String> properties, boolean isOverwrite, boolean isDelete,
+        String deleteKey, final SetPropertiesCallback callback, Object ctx) {
+        if (!metadataMutex.tryLock()) {
+            // Defer update for later
+            scheduledExecutor.schedule(() -> asyncUpdateProperties(properties, isOverwrite, isDelete, deleteKey,
+                callback, ctx), 100, TimeUnit.MILLISECONDS);
+            return;
+        }
+        if (isDelete) {
+            propertiesMap.remove(deleteKey);
+        } else if (isOverwrite) {
```

Comment:
I still think the `setProperties()` should not blindly overwrite the current map, rather just overwrite the individual keys. It's just a convenience to have 1 metadata op to set multiple keys.

Otherwise it would be a `replaceProperties()`, but I don't see how would that be more useful than a `setProperties()`.

Reply:
Done.